File: pyqt/info_dialog.py
# !/usr/bin/python3
# -*- coding: utf-8 -*-
import os
import sys
import json
from pathlib import Path
from PySide6.QtWidgets import QApplication, QWidget, QMessageBox
from PySide6.QtCore import Slot, Signal
from shiboken6.Shiboken import invalidate

import global_config
from ui.ui_info_dialog import Ui_InfoDialog

class InfoDialog(QWidget):
    # custom signals
    begin_setting = Signal(str)     # call the start_panel 

    def __init__(self):
        super(InfoDialog, self).__init__()
        self.ui = Ui_InfoDialog()
        self.ui.setupUi(self)
        self.setWindowTitle('实验参与者信息填写')
        
        self.is_debug = global_config.get_value('mode', 'debug', False)
        self.out_data_path = Path(global_config.get_value('data', 'path', './outdata')).absolute()
        self.participant_id = 'debug'
        self.info_path = os.path.join(self.out_data_path,self.participant_id)
        self.participant_info_file = os.path.join(self.info_path,'participant_info.json')
        self.info_data = {}
        
        self.setup_connections()


    def setup_connections(self):
        pass

    def check_info(self):
        # get input information
        info_data = {}
        info_data['name'] = self.ui.tEdit_name.text()
        info_data['age'] = self.ui.tEdit_age.text()
        info_data['id'] = self.ui.tEdit_id.text()
        info_data['major'] = self.ui.tEdit_major.text()
        if self.ui.btn_man.isChecked():
            info_data['sex'] = '男'
        elif self.ui.btn_woman.isChecked():
            info_data['sex'] = '女'
            
        # check if information is null   
        invalid_fields = []
        for field in ('name', 'age', 'id'):
            if info_data[field].strip() == '':
                invalid_fields.append(field)
        return invalid_fields,info_data
    
            
    # custom slots
    @Slot(bool)
    def info_show(self):
        self.show()
 
    @Slot()
    def on_btn_submit_clicked(self):
        invalid_fields,info_data = self.check_info()
        if len(invalid_fields):
            if self.is_debug:
                msg_title = '[Debug Mode]'
                invalid_fields_str = ''
                for field in invalid_fields:
                    invalid_fields_str = invalid_fields_str + field + ','
                msg_text = invalid_fields_str + "is invalid! id is set to 'debug'"
                QMessageBox.warning(self, msg_title, msg_text)
                info_data['id'] = self.participant_id
            else:
                msg_title = '信息填写错误'
                field_dict = {'name': '姓名', 'age': '年龄', 'id': '学号/编号'}
                invalid_fields_str = ''
                for field in invalid_fields:
                    invalid_fields_str = invalid_fields_str + field_dict[field] + ' '
                msg_text = invalid_fields_str + '未填写！'
                QMessageBox.warning(self, msg_title, msg_text)
                return
        submit_choose = QMessageBox.question(self, '信息提交确认', '信息填写完成，是否确认提交？',
                                             QMessageBox.Yes | QMessageBox.Cancel,
                                             QMessageBox.Yes)
        if submit_choose == QMessageBox.Yes:
            self.info_data = info_data
            self.participant_id = info_data['id']
            self.info_path = os.path.join(self.out_data_path, self.participant_id)
            if not os.path.exists(self.info_path):
                os.makedirs(self.info_path)
                self.terminate()
            else:
                cover = QMessageBox.question(self, '提示', '该编号已存在，是否覆盖？\n 覆盖(Yes) 自动重编号(No) 修改编号(Cancel)',
                                             QMessageBox.Yes | QMessageBox.No | QMessageBox.Cancel,
                                             QMessageBox.No)
                if cover == QMessageBox.Yes:
                    self.terminate()
                elif cover == QMessageBox.No:
                    repeat_no = 1
                    repeat_str = '_rep' + str(repeat_no)
                    self.info_path = os.path.join(self.out_data_path, info_data['id'] + repeat_str)
                    while os.path.exists(self.info_path):
                        repeat_no = repeat_no + 1
                        repeat_str = '_rep' + str(repeat_no)
                        self.info_path = os.path.join(self.out_data_path, info_data['id'] + repeat_str)
                    os.makedirs(self.info_path)
                    self.terminate()
                else:
                    QMessageBox.information(self, '取消覆盖', '你可以修改编号后再提交')
                    return
        else:
            QMessageBox.information(self, '取消提交', '你可以继续填写信息')
            return

    # normal methods
    def terminate(self):
        self.participant_info_file = os.path.join(self.info_path, 'participant_info.json')
        fid = open(self.participant_info_file, mode='w+')
        json.dump(self.info_data, fid, indent=4, ensure_ascii=False)
        fid.close()
        self.begin_setting.emit(self.participant_id)
        self.close()


if __name__ == "__main__":
    global_config.init()
    app = QApplication([])
    widget = InfoDialog()
    widget.show()
    sys.exit(app.exec_())

File: cpp/src/infodialog.cpp
#include "infodialog.h"
#include "ui_infodialog.h"

#include <QJsonDocument>
#include <QJsonObject>
#include <QMessageBox>
#include <QtDebug>

InfoDialog::InfoDialog(QWidget *parent)
    : QDialog(parent), ui(new Ui::InfoDialog)
{
    ui->setupUi(this);
    this->setWindowTitle(tr("实验参与者信息填写"));
    this->is_debug = global_config.get_value("mode/debug", false).toBool();
    this->out_data_path = global_config.get_value("data/path", "./outdata").toString();
    this->participant_id = tr("debug");
    this->info_path = this->out_data_path.absoluteFilePath(this->participant_id);
    this->info_data = {{"name", ""}, {"age", ""}, {"id", ""}, {"major", ""}, {"sex", ""}};

    this->setup_connections();
}

InfoDialog::~InfoDialog()
{
    delete ui;
}

void InfoDialog::setup_connections(){};

QStringList InfoDialog::check_info()
{
    info_data["name"] = this->ui->tEdit_name->text();
    info_data["age"] = this->ui->tEdit_age->text();
    info_data["id"] = this->ui->tEdit_id->text();
    info_data["major"] = this->ui->tEdit_major->text();
    if (this->ui->btn_man->isChecked())
        info_data["sex"] = tr("男");
    else if (this->ui->btn_woman->isChecked())
        info_data["sex"] = tr("女");
    QStringList invalid_fields;
    QStringList fields_to_check = {"name", "age", "id"};
    for (QString &field : fields_to_check)
    {
        if (info_data.contains(field))
        {
            QString value = info_data[field].toString().trimmed();
            if (value.size() == 0)
                invalid_fields.append(field);
        }
        else
        {
            qCritical() << tr("can't find field %1 in QHash info_data").arg(field);
        }
    }
    return invalid_fields;
};

void InfoDialog::terminate()
{
    this->participant_info_file.setFileName(this->info_path.absoluteFilePath("participant_info.json"));
    QJsonObject info_obj = QJsonObject::fromVariantHash(this->info_data);
    QJsonDocument info_doc(info_obj);
    bool ok = this->participant_info_file.open(QIODevice::WriteOnly);
    if (ok)
    {
        this->participant_info_file.write(info_doc.toJson());
        this->participant_info_file.close();
        qInfo() << "participant info json file write success!";
    }
    else
    {
        qDebug() << "participant info json file write error!";
    }
    emit begin_setting(this->participant_id);
    this->close();
};

void InfoDialog::on_btn_submit_clicked()
{
    QStringList invalid_fields = this->check_info();
    if (invalid_fields.size())
    {
        QString msg_title = tr("信息填写错误");
        QHash<QString, QString> field_dict = {{"name", tr("姓名")}, {"age", tr("年龄")}, {"id", tr("学号/编号")}};
        QString invalid_fields_str;
        for (QString &field : invalid_fields)
        {
            invalid_fields_str.append(field_dict[field]).append(" ");
        }
        invalid_fields_str.append(tr("未填写!"));
        QMessageBox::warning(this, msg_title, invalid_fields_str);
        if (!this->is_debug)
            return;
        else
        {
            QString msg_text = invalid_fields_str.append("is invalid! id is set to 'debug'");
            QMessageBox::warning(this, msg_title, msg_text);
            this->info_data["id"] = this->participant_id;
            this->terminate();
            return;
        }
    }
    QMessageBox::StandardButton submit_choose = QMessageBox::question(this,
                                                                      tr("信息提交确认"),
                                                                      tr("信息填写完成,是否确认提交？"),
                                                                      QMessageBox::Yes | QMessageBox::Cancel,
                                                                      QMessageBox::Yes);
    if (submit_choose == QMessageBox::Yes)
    {
        this->participant_id = this->info_data["id"].toString();
        this->info_path = this->out_data_path.absoluteFilePath(this->participant_id);
        if (!this->info_path.exists())
        {
            this->out_data_path.mkpath(this->participant_id);
            this->terminate();
        }
        else
        {
            QMessageBox::StandardButton is_cover = QMessageBox::question(this,
                                                                         tr("提示"),
                                                                         tr("该编号已存在，是否覆盖？ \n"
                                                                            "覆盖(Yes) 自动重编号(No) 手动修改编号(Cancel)"),
                                                                         QMessageBox::Yes | QMessageBox::No | QMessageBox::Cancel,
                                                                         QMessageBox::No);
            if (is_cover == QMessageBox::Yes)
                this->terminate();
            else if (is_cover == QMessageBox::No)
            {
                int repeat_no = 1;
                QString repeat_str = QString("_rep%1").arg(repeat_no);
                this->info_path = this->out_data_path.absoluteFilePath(this->participant_id + repeat_str);
                while (this->info_path.exists())
                {
                    repeat_no += 1;
                    repeat_str = QString("_rep%1").arg(repeat_no);
                    this->info_path = this->out_data_path.absoluteFilePath(this->participant_id + repeat_str);
                }
                this->participant_id += repeat_str;
                this->info_path.mkpath(this->info_path.absolutePath());
                this->terminate();
            } 
            else
            {
                QMessageBox::information(this, tr("取消覆盖"), tr("你可以修改编号后再提交"));
                return;
            }
        }
    }
    else
    {
        QMessageBox::information(this, tr("取消提交"), tr("你可以继续填写信息"));
        return;
    }
};
